Introduce Mining interface

Sjors opened the pull request:
Introduce a `Mining` interface for the `getblocktemplate`, `generateblock` and other mining RPCs to use now, and for Stratum v2 to use later.

Suggested here: https://github.com/bitcoin/bitcoin/pull/29346#issuecomment-2108528652

The selection of methods added to the interface is mostly based on what the Template Provider in #29432 uses. It could be expanded further so that `rpc/mining.cpp` no longer needs `EnsureMemPool` and `EnsureChainman`.

This PR should be a pure refactor.

Sjors commented:
The commit is all the boilerplate plus a single method `isTestChain()` that's used by the `getblocktemplate` RPC.

Sjors committed:
rpc: getblocktemplate getTipHash() via Miner interface

Sjors committed:
rpc: getTransactionsUpdated via miner interface

Sjors committed:
rpc: call processNewBlock via miner interface

Sjors commented:
Let me know if I should expand the interface to cover more of `rpc/mining.cpp` or stick to this for now.

Sjors commented on src/node/interfaces.cpp:
```diff
@@ -854,6 +856,13 @@ class MinerImpl : public Miner
+    std::unique_ptr<CBlockTemplate> createNewBlock(const CScript& scriptPubKeyIn) override
+    {
+        LOCK(::cs_main);
+        // TODO pass mempool sanely
```
f04cf9f75054573c204245f7bb1e6813cce1fed8: I guess I should change the `BlockAssembler` constructor

Sjors commented on src/node/interfaces.cpp:
```diff
@@ -854,6 +856,13 @@ class MinerImpl : public Miner
+    {
+        LOCK(::cs_main);
+        // TODO pass mempool sanely
+        return BlockAssembler{chainman().ActiveChainstate(), &*(context()->mempool)}.CreateNewBlock(scriptPubKeyIn);
```
f04cf9f75054573c204245f7bb1e6813cce1fed8: I guess I should change the `BlockAssembler` constructor

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/30200#discussion_r1621067334

> I guess I should change the `BlockAssembler` constructor

In commit "rpc: call CreateNewBlock via miner interface" (f04cf9f75054573c204245f7bb1e6813cce1fed8)

I don't actually understand the TODO here. The way the mempool is passed seems ok. You could do mempool.get() instead of &*mempool, but both should be equivalent.

Sjors replied:
Thanks, replaced it with `get()`.

Sjors commented on src/rpc/server_util.h:
```diff
@@ -31,6 +34,7 @@ ChainstateManager& EnsureAnyChainman(const std::any& context);
 CBlockPolicyEstimator& EnsureFeeEstimator(const node::NodeContext& node);
 CBlockPolicyEstimator& EnsureAnyFeeEstimator(const std::any& context);
 CConnman& EnsureConnman(const node::NodeContext& node);
+interfaces::Miner& EnsureMiner(const node::NodeContext& node);
```
@ryanofsky I'm not sure what the design philophy is behind `Ensure` and `EnsureAny`. The other methods here return more low level objects.

Perhaps related to what you said here:

> > I think of "Context" as being a passive container for state instead of an object you would call methods on. The top level interfaces like `interfaces::Chain` `interface::Node` `interfaces::Wallet` are actually stateless themselves, and only point to state stored in other places. (The idea is to let interface instances be created and destroyed freely, since with multiprocess support new processes can connect and create interface pointers that get destroyed on disconnection, and the state of the node shouldn't be affected too much from connections and disconnections. Anyway, the mining interface could follow this pattern.)

TheCharlatan commented:
Concept ACK

ryanofsky commented on src/interfaces/miner.h:
```diff
@@ -0,0 +1,66 @@
+//! Interface giving clients (RPC, Stratum v2 Template Provider in the future)
+//! ability to create block templates.
+
+class Miner
```
In commit "Introduce Miner interface" (b30a8aca5191186f56eb1e9bd8f1a1d899a07039)

Not important, but could consider renaming this class.

- interfaces::Node is an interface used by the gui to control the node
- interfaces::Wallet is an interface used by the gui to control the node
- interface::Chain is an interface used by the wallet and indexes to sync with the chain

interfaces::Miner sounds like an interface that would be used to control a miner, which is not really what this is. Maybe interfaces::Mining or interfaces::MiningService would be better.

Sjors replied:
Renamed to `Mining`

ryanofsky commented on src/rpc/mining.cpp:
```diff
@@ -694,6 +696,7 @@ static RPCHelpMan getblocktemplate()
                 throw JSONRPCError(RPC_DESERIALIZATION_ERROR, "Block decode failed");
 
             uint256 hash = block.GetHash();
+
```
In commit "rpc: call TestBlockValidity via miner interface" (d880aa0ebe374394b2f8c061ebc7d428f899b59b)

Maybe unintended whitespace change here

ryanofsky commented on src/rpc/mining.cpp:
```diff
@@ -761,7 +759,7 @@ static RPCHelpMan getblocktemplate()
         {
             // NOTE: Spec does not specify behaviour for non-string longpollid, but this makes testing easier
-            hashWatchedChain = active_chain.Tip()->GetBlockHash();
+            hashWatchedChain = miner.getTip()->GetBlockHash();
```
In commit "rpc: getblocktemplate getTip() via Miner interface" (d5b354ed93266d76ab9e2dd2af9c9b611583a1ee)

Would suggest introducing a `CBlockIndex* tip{miner.getTip()};` variable, using it where possible, and updating it as needed instead of calling `miner.getTip()` repeatedly in this function.

It seems like the reason it is ok to make multiple getTip() calls is that cs_main is locked for most of this function. But more ideally, this function would not be locking cs_main directly, only calling interface methods which lock cs_main internally. Using a tip variable in this code, instead of just assuming tip cannot change, would get closer to this state.

Sjors replied:
I'm now setting `uint256 tip{miner.getTipHash()};` at the top. For long polls it's updated after the wait. See d1be135e568605539fa3932da6e7c5032816825b.

ryanofsky reviewed (approved):
Code review ACK 87ba70a2731951a16432ca2d8d55c62bad87dd41

re: https://github.com/bitcoin/bitcoin/pull/30200#issuecomment-2140187315

> Let me know if I should expand the interface to cover more of `rpc/mining.cpp` or stick to this for now.

I think it's just important for the interface to be usable for stratum code so is not too tied to node internals. It should be ok for RPC code to go around the interface and not use it for everything.

Sjors committed:
Introduce Mining interface

Start out with a single method isTestChain() that's used by the getblocktemplate RPC.

Sjors committed:
rpc: minize getTipHash() calls in gbt

Set tip at the start of the function and only update it for a long poll.

Additionally have getTipHash return an optional, so the
caller can explicitly check that a tip exists.

Sjors commented on src/interfaces/miner.h:
```diff
@@ -0,0 +1,66 @@
+    virtual bool isTestChain() = 0;
+
+    //! Returns the index entry for the tip of this chain, or nullptr if none.
+    virtual CBlockIndex* getTip() = 0;
```
I replaced `getTip()` with `getTipHash()`. `Sv2TemplateProvider` doesn't need it, so for now I use it in a few places in `rpc/mining.cpp`, but then fall back to `chainman.m_blockman.LookupBlockIndex(miner.getTipHash())` where we need more info.

A future refactor could move more of `getblocktemplate` implementation details out of the RPC.

Sjors commented:
@ryanofsky thanks for the review! I rebased and changed a few things to address comments, see inline.

Also added `IsInitialBlockDownload`, which was the last thing `Sv2TemplateProvider` still needed `ChainstateManager& m_chainman;` for. 

I'll cherry-pick this into #29432 to see if we need anything more.

Sjors commented:
All the above should be fixed now, except for `g_best_block_cv.wait_until` (but that's not blocking).

The biggest change is in 914c26bbd4d7e87af4e0d0edf9af69ae76668b7e. It now makes the `options` argument mandatory for `BlockAssembler` constructor, dropping implicit handling of `ArgsManager`. The caller i.e. the `Mining` interface implementation now handles this.

In Stratum v2 the pool communicates how many extra bytes it needs for its own outputs (payouts, extra commitments, etc). This needs to be subtracted from what the user set as -blockmaxweight. To achieve that the caller would have to pass in an options object, and not forget to also process `-blockmintxfee` - a mistake I made in #29432.

Sjors commented:
You can see the interface in action in https://github.com/bitcoin/bitcoin/pull/29432/commits/1618c3dd86d65cf00f5f3c2c30c1db2b0e9dd9ca

Sjors commented:
The `generateblock` block RPC relies on not passing the mempool to `BlockAssembler`, in order to allow manual transaction selection. So I added `bool use_mempool = true` to `createNewBlock`.

ryanofsky commented on src/node/miner.h:
```diff
@@ -203,9 +202,6 @@ int64_t UpdateTime(CBlockHeader* pblock, const Consensus::Params& consensusParam
 void RegenerateCommitments(CBlock& block, ChainstateManager& chainman);
-
-/** Apply -blockmintxfee and -blockmaxweight options from ArgsManager to BlockAssembler options. */
-void ApplyArgsManOptions(const ArgsManager& gArgs, BlockAssembler::Options& options);
```
In commit "rpc: call CreateNewBlock via miner interface" (f375fb9a0dff7ccead80626f4aee853db7799e6f)

It might be good to keep this ApplyArgsManOptions function for more consistency with other options parsing code. The node/interfaces.cpp createNewBlock function could call ApplyArgsManOptions instead of parsing options itself. The current approach also seems ok though if you prefer it.

Sjors replied:
I split f375fb9a0dff7ccead80626f4aee853db7799e6f in two parts:

1. 6d6ddcb759465c7375ff772faa960cf92805468a simply adds and uses the interface
2. 6b7c8c3e27ae529306d18019e7bd94b4ae5b9a21 drops the `BlockAssembler` constructor without `options` argument

I changed this second commit to keep `ApplyArgsManOptions`.

ryanofsky commented on src/node/interfaces.cpp:
```diff
@@ -856,6 +859,20 @@ class MinerImpl : public Mining
+        BlockAssembler::Options options;
+
+        // Block resource limits
+        options.nBlockMaxWeight = gArgs.GetIntArg("-blockmaxweight", options.nBlockMaxWeight);
+        if (const auto blockmintxfee{gArgs.GetArg("-blockmintxfee")}) {
+            if (const auto parsed{ParseMoney(*blockmintxfee)}) options.blockMinFeeRate = CFeeRate{*parsed};
+        }
```
In commit "rpc: call CreateNewBlock via miner interface" (f375fb9a0dff7ccead80626f4aee853db7799e6f)

I'm not clear how this implementation lines up with the commit description which says "the pool communicates how many extra bytes it needs for its own outputs (payouts, extra commitments, etc). This needs to be substracted from what the user set as -blockmaxweight."

Is the commit description just saying that this implementation is not complete and it will need to be extended in the future? Would be helpful to clarify.

Sjors replied:
Yes, this is a reference to the future change in sv2. I'll see if I can reword it.

ryanofsky reviewed (approved):
Code review ACK febd3d1c3d41cf6083baf990fb2a3a69cade364a

Sjors committed:
rpc: call CreateNewBlock via miner interface

Sjors committed:
Always pass options to BlockAssembler constructor

This makes the options argument for BlockAssembler constructor mandatory,
dropping implicit use of ArgsManager. The caller i.e. the Mining
interface implementation now handles this.

In a future Stratum v2 change the Options object needs to be
mofified after arguments have been processed. Specifically
the pool communicates how many extra bytes it needs for
its own outputs (payouts, extra commitments, etc). This will need
to be substracted from what the user set as -blockmaxweight.

Such a change can be implemented in createNewBlock, after
ApplyArgsManOptions.

Sjors committed:
rpc: call TestBlockValidity via miner interface

Sjors commented on src/interfaces/mining.h:
```diff
@@ -25,6 +26,9 @@ class Mining
     virtual bool isTestChain() = 0;
 
+    //! Returns the hash for the tip of this chain, 0 if none
+    virtual uint256 getTipHash() = 0;
```
Done

Sjors commented on src/rpc/mining.cpp:
```diff
@@ -703,12 +706,14 @@ static RPCHelpMan getblocktemplate()
             }
 
+            BlockValidationState state;
+            miner.testBlockValidity(state, block, /*check_merkle_root=*/true);
```
I think that was by accident. Moved it back.

Sjors commented:
@ryanofsky do you have any thoughts on this (for this PR or a followup)?

> Additionally the template provider needs to wait, to see if there's a new tip, with g_best_block_cv.wait_until. How would I go about moving that into this interface?

ryanofsky commented:
> > Additionally the template provider needs to wait, to see if there's a new tip, with g_best_block_cv.wait_until. How would I go about moving that into this interface?

Maybe the simplest way to do it would be to have a waitTipChanged() method that just blocks waiting for a new tip, and would be interrupted when the node shuts down or the mining interface is destroyed. If needed, the waiting interface could have more features and accept a timeout or return an object with wait() and cancel() methods.

Another alternative would be to provide a callback interface instead of a waiting interface like interfaces::Node:

https://github.com/bitcoin/bitcoin/blob/0fbb8043ce0d6a931d5ce4bc877b6b596976a908/src/interfaces/node.h#L256-L259

ryanofsky reviewed (approved):
Code review ACK 4f9c3369a453f7e5f0c12179dd629ee8fc042379. Only changes since last review are: restoring ApplyArgsManOptions function and inconclusive-not-best-prevblk check order, splitting up an internal commit, and improving some commit messages.

Sjors commented:
Thanks for the review, and for the suggestions. I might add that later to this PR or as a followup.

Trivial rebase after #29015.

ryanofsky reviewed (approved):
Code review ACK 5666af2c6e4924bf8e358ab5121c88d14cb085df, just rebased since last review

ismaelsadeeq reviewed (commented):
Concept ACK

tdb3 reviewed (approved):
ACK 5666af2c6e4924bf8e358ab5121c88d14cb085df
Great job laying groundwork for Sv2.

Left some comments (mostly nits).  Ran unit and functional tests (passed).

Other than running unit/functional tests and exercising test networks (e.g. regtest for `generate` calls), any other good ways you'd like to see these changes exercised `getblocktemplate()` is a pretty crucial RPC, so being extra paranoid here could be prudent.

p.s. May want to update the description, since RPC functions beyond `getblocktemplate()` and `gererateblock()` also use the new Mining interface (e.g. `submitblock()`, `generatetodescriptor()`, `generatetoaddress()`, etc.)

Sjors commented:
> any other good ways you'd like to see these changes exercised `getblocktemplate()` is a pretty crucial RPC

You could test #27433 on top of the changes here. Or try solo CPU mining on a custom signet, e.g. using [pooler/cpuminer](https://github.com/pooler/cpuminer) and then:

```sh
./minerd -u bitcoin -p bitcoin -o http://127.0.0.1:38332 --coinbase-addr $ADDRESS --no-stratum --algo=sha256d
```

The latter requires `signetchallenge=51` and a patch setting `#define GBT_RULES "[\"segwit\", \"signet\"]"`.

Sjors commented on src/node/interfaces.cpp:
```diff
@@ -844,6 +844,14 @@ class MinerImpl : public Mining
+        LOCK(::cs_main);
+        CBlockIndex* tip{chainman().ActiveChain().Tip()};
+        if (!tip) return uint256{0};
+        return tip->GetBlockHash();
```
`GetBlockHash()`, which was called directly before this commit, has an `assert` that `tip` must exist. So `return uint256{0};` should never happen. An optional does seem more robust though, will look into it.

Sjors commented on src/interfaces/mining.h:
```diff
@@ -22,6 +25,13 @@ class Mining
     /** If this chain is exclusively used for testing */
     virtual bool isTestChain() = 0;
 
+    /**
+     * Check a block is completely valid from start to finish.
+     * Only works on top of our current best block.
+     * Does not check proof-of-work.
+     * */
+    virtual bool testBlockValidity(BlockValidationState& state, const CBlock& block, bool check_merkle_root) = 0;
```
I added a default `true` to mimic `TestBlockValidity` in `validation.h`.

Sjors commented:
Oops, `'TestBlockValidity failed: bad-txns-inputs-missingorspent'`...

Sjors commented:
Rebased and added some missing `#include`s and `class`es.

Sjors committed:
rpc: call IsInitialBlockDownload via miner interface

tdb3 reviewed (approved):
re ACK a9716c53f05082d6d89ebea51a46d4404efb12d7

Re-built and ran unit/functional tests (passed).
Tested https://github.com/bitcoin/bitcoin/pull/30200#issuecomment-2172702822 by mining with cpuminer (https://github.com/pooler/cpuminer) for a custom signet (with `signetchallenge=51` set in bitcoin.conf to allow mining without signature).

Mined for 5000 blocks, and created and spent several transactions (one from 135 coinbase outputs and two derivative transactions).  Temporarily paused mining, created and broadcasted three additional transactions.  Called getblocktemplate to verify that the transactions were included (`src/bitcoin-cli -datadir=/home/dev/customsignet getblocktemplate '{"rules":["segwit","signet"]}'`), then continued mining to confirm the transactions.  All mining and spending was successful.

This Issue from the cpuminer repo was helpful in bootstrapping (thanks @sjors): https://github.com/pooler/cpuminer/issues/285

Useful conf and cli:

bitcoin.conf:
```
signet=1
rpcuser=bitcoin
rpcpassword=bitcoin
[signet]
signetchallenge=51
```

```
src/bitcoin-cli -datadir=/home/dev/customsignet getblocktemplate '{"rules":["segwit","signet"]}'
```

cpuminer:
```
./minerd -u bitcoin -p bitcoin -o http://127.0.0.1:38332 --coinbase-addr tb1qwz06m34zc9e5cy8tjz7f3x5xemega5rn0waq0v --no-stratum --algo=sha256d
```

itornaza reviewed (approved):
Code review and std-tests ACK a9716c53f05082d6d89ebea51a46d4404efb12d7

Redirected to this PR from https://github.com/bitcoin/bitcoin/pull/29346#issuecomment-2139460402. I really like how clearly the mining interface is layed out after @ryanofsky comment https://github.com/bitcoin/bitcoin/pull/29346#issuecomment-2108528652 to further support @Sjors work on integrating the noise protocol that seems to be needed for stratum v2.

I have closely followed the commit history, built and run all standard unit, functional and extended tests, and all of them pass.

Since `src/interface/mining.h` is a new source file, I included a couple of non-blocking styling nits.

ryanofsky commented on src/interfaces/mining.h:
```diff
@@ -22,6 +25,18 @@ class Mining
+     * @param[in] check_merkle_root call CheckMerkleRoot()
+     * @returns false if any of the checks fail
+     */
+    virtual bool testBlockValidity(BlockValidationState& state, const CBlock& block, bool check_merkle_root = true) = 0;
```
In commit "rpc: call TestBlockValidity via miner interface" (d8a3496b5ad27bea4c79ea0344f595cc1b95f0d3)

Not important, but it might be good to move the `state` output parameter last. The reason would be to work with the libmultiprocess code generator in case we want to expose this interface over IPC. The code generator only knows how to parse Cap'n Proto declarations, not C++ code, so when it encounters a Cap'n Proto method declaration with output parameters like:

```capnp
testBlockValidity @1 (block: Data, maxTxFee :Bool) -> (state: ValidationState, result :Bool);
```

it just assumes a corresponding C++ method exists with the listed input parameters first and output parameters last.

(Also I just noticed information about output parameters was changed in the developer notes. Suggestion to put output parameters last used used to be mentioned in the "internal interface" style section, but it looks like e66b321fd1ddfffd9bfc59d407ad8f03490b873c from #25092 turned that advise into a bigger, broader rule, which I'm not sure is a great change..)

Sjors replied:
Thanks, I'll change this if I have to retouch.

Sjors replied:
I made a note to do this along with a PR to add something like `waitTipChanged()` 
https://github.com/bitcoin/bitcoin/pull/30200#issuecomment-2160764070

Sjors replied:
Actually, doing this now since I ran into another (minor) issue: #30335

ryanofsky reviewed (commented):
Code review ACK a9716c53f05082d6d89ebea51a46d4404efb12d7 with one minor suggestion in case you update. Only changes since last review were other small changes to the interface.

Sjors commented on src/interfaces/mining.h:
```diff
@@ -0,0 +1,82 @@
+     * @param[in] use_mempool set false to omit mempool transactions
+     * @returns a block template
+     */
+    virtual std::unique_ptr<node::CBlockTemplate> createNewBlock(const CScript& script_pub_key, bool use_mempool = true) = 0;
```
Fixed in #30335

AngusP commented:
(late) Code Review ACK a9716c53f05082d6d89ebea51a46d4404efb12d7

Also this was good to read to get more familair with the code touched, thanks!

TheCharlatan commented on src/node/interfaces.cpp:
```diff
@@ -859,6 +860,12 @@ class MinerImpl : public Mining
 
+    std::unique_ptr<CBlockTemplate> createNewBlock(const CScript& script_pub_key, bool use_mempool) override
+    {
+        LOCK(::cs_main);
```
Is this lock required?

Sjors replied:
Probably not, because `BlockAssembler::CreateNewBlock` takes a lock itself. As does `ChainstateManager::ActiveChainstate()`.

I think I initially added this because `generateblock()` in `rpc/mining.cpp` did it.

I'll add a commit to #30335 to drop it.

@maflcko found something similar

Sjors replied:
(unsure if that was indeed the reason)

ryanofsky commented:
> Is it ok for other external users of `processNewBlock` to skip over the pre-checks and optimizations that are normally done in `submitblock`?

I assumed it was ok, but now that `interfaces::BlockTemplate::submitSolution` in added in 525e9dcba0b8c6744bcd3725864f39786afc8ed5 (to deal with the concern in https://github.com/bitcoin/bitcoin/pull/30440#issuecomment-2229233817) maybe the `interfaces::Mining::processNewBlock` method could be dropped and RPC code can go back to calling `ProcessNewBlock` directly (effectively reverting 7b4d3249ced93ec5986500e43b324005ed89502f) if that is preferred.

I don't think it matters much to the RPC implementation which function is called. The reason for switching RPCs to use the mining interface is to test the interface and to take advantage of new features and conveniences it adds, which do not seem to be present in this case.

Sjors commented:
I opened #31196 to drop `processNewBlock`.
